rpcq/core-messages.py

```diff
@@ -44,37 +44,183 @@ class Resources(Message):
 
 
 @dataclass(eq=False, repr=False)
-class Waveform(Message):
+class AbstractWaveform(Message):
     """
-    A waveform envelope defined for a specific frame.
+    A waveform envelope defined for a specific frame. This abstract class is made concrete by either a `Waveform` or a templated waveform such as `GaussianWaveform` 
     """
 
     frame: str
     """The label of the associated tx-frame."""
 
+
+@dataclass(eq=False, repr=False)
+class Waveform(AbstractWaveform):
+    """
+    A waveform envelope defined by specific IQ values for a specific frame.
+    """
+
     iqs: List[float] = field(default_factory=list)
-    """The raw waveform envelope samples,
-          alternating I and Q values."""
+    """The raw waveform envelope samples, alternating I and Q values."""
+
+
+@dataclass(eq=False, repr=False)
+class TemplateWaveform(AbstractWaveform):
+    """
+    A waveform envelope defined for a specific frame. A templated waveform is defined by a parameterized pulseshape rather than explicit IQ values. The message specification does not enforce that the duration is implementable on the hardware.
+    """
+
+    duration: float
+    """Length of the pulse in seconds"""
+
+    scale: float = 1.e+0
+    """Scale to apply to waveform envelope"""
+
+    phase: float = 0.0e+0
+    """Phase [units of tau=2pi] to rotate the complex waveform envelope."""
+
+    detuning: float = 0.0e+0
+    """Modulation to apply to the waveform in Hz"""
+
+
+@dataclass(eq=False, repr=False)
+class GaussianWaveform(TemplateWaveform):
+    """
+    A Gaussian shaped waveform envelope defined for a specific frame.
+    """
+
+    fwhm: Optional[float] = None
+    """Full Width Half Max shape paramter in seconds"""
+
+    t0: Optional[float] = None
+    """Center time coordinate of the shape in seconds. Defaults to mid-point of pulse."""
+
+
+@dataclass(eq=False, repr=False)
+class DragGaussianWaveform(TemplateWaveform):
+    """
+    A Gaussian shaped waveform envelope defined for a specific frame.
```

Comment:
This docstring should probably be different than the docstring for GaussianWaveform.